builtin/credential/aws/path_login.go

```diff
@@ -888,6 +1048,389 @@ func (b *backend) pathLoginRenew(
+	}
+
+	endpoint := "https://sts.amazonaws.com"
+
+	if config != nil {
+		if config.IAMServerIdHeaderValue != "" {
+			err = validateVaultHeaderValue(headers, parsedUrl, config.IAMServerIdHeaderValue)
+			if err != nil {
+				return logical.ErrorResponse(fmt.Sprintf("error validating %s header: %s", iamServerIdHeader, err)), nil
+			}
+		}
+		if config.STSEndpoint != "" {
+			endpoint = config.Endpoint
+		}
+	}
+
+	clientArn, accountID, err := submitCallerIdentityRequest(method, endpoint, parsedUrl, body, headers)
+	if err != nil {
+		return logical.ErrorResponse(fmt.Sprintf("error making upstream request: %s", err)), nil
+	}
+	canonicalArn, principalName, sessionName, err := parseIamArn(clientArn)
+	if err != nil {
+		return logical.ErrorResponse(fmt.Sprintf("Error parsing arn: %s", err)), nil
+	}
+
+	roleName := data.Get("role").(string)
+	if roleName == "" {
+		roleName = principalName
+	}
+
+	roleEntry, err := b.lockedAWSRole(req.Storage, roleName)
+	if err != nil {
+		return nil, err
+	}
+	if roleEntry == nil {
+		return logical.ErrorResponse(fmt.Sprintf("entry for role %s not found", roleName)), nil
+	}
+
+	if !strutil.StrListContains(roleEntry.AllowedAuthTypes, iamAuthType) {
+		return logical.ErrorResponse(fmt.Sprintf("auth method iam not allowed for role %s", roleName)), nil
+	}
+
+	// The role creation should ensure that either we're inferring this is an EC2 instance
+	// or that we're binding an ARN
+	if roleEntry.BoundIamPrincipalARN != "" && roleEntry.BoundIamPrincipalARN != canonicalArn {
+		return logical.ErrorResponse(fmt.Sprintf("IAM Principal %q does not belong to the role %q", clientArn, roleName)), nil
+	}
+
+	policies := roleEntry.Policies
+	rTagMaxTTL := time.Duration(0)
+
+	inferredEntityType := ""
+	inferredEntityId := ""
+	if roleEntry.RoleInferredType == ec2EntityType {
+		instance, err := b.validateInstance(req.Storage, sessionName, roleEntry.InferredAWSRegion, accountID)
+		if err != nil {
+			return logical.ErrorResponse(fmt.Sprintf("failed to verify %s as a valid EC2 instance in region %s", sessionName, roleEntry.InferredAWSRegion)), nil
+		}
+
+		// build a fake identity doc to pass on metadata about the instance to verifyInstanceMeetsRoleRequirements
+		identityDoc := &identityDocument{
+			Tags:        nil, // Don't really need the tags, so not doing the work of converting them from Instance.Tags to identityDocument.Tags
+			InstanceID:  *instance.InstanceId,
+			AmiID:       *instance.ImageId,
+			AccountID:   accountID,
+			Region:      roleEntry.InferredAWSRegion,
+			PendingTime: instance.LaunchTime.Format(time.RFC3339),
+		}
+
+		roleTagResp, validationError, err := b.verifyInstanceMeetsRoleRequirements(req.Storage, instance, roleEntry, roleName, identityDoc)
+		if err != nil {
+			return nil, err
+		}
+		if validationError != nil {
+			return logical.ErrorResponse(fmt.Sprintf("Error validating instance: %s", validationError)), nil
+		}
+
+		if roleTagResp != nil {
```

Comment:
Apologies, but I'm struggling to fully grok what is being suggested here. Just for my own sanity, is the crux of the issue here that role tags contain fields specific to both instances and the auth type used by instances?

@vishalnayak -- your number 2 and 3 at the end were pretty much exactly as I was holding it; however, the way I saw it, role tags are things that apply to instances; they're not specific to the instance identity document. Therefore, I tried to keep most of the things in the role tag that made sense to apply (`disallow_reauthentication` and `allow_instance_migration` being the two that I took out).

My view is that how I've done it provides a smooth transition path for customers from the `ec2` auth method to the `iam` auth method. They can enable `iam` auth on the role and configure inferencing, and then gradually transition clients over to use `iam` and then turn off `ec2` once they've migrated. To facilitate this migration, I would want the behavior to be as close as possible between the `ec2` and the `iam` auth methods, which means I wouldn't want customers surprised to find out that their instances suddenly were able to assume a broader set of policies than they intended because we started ignoring role tags. Keeping role tags should also have the side benefit of cutting down on role sprawl inside of Vault. Of course, this migration path I mentioned would require the ability to have `ec2` and `iam` enabled at the same time.

> For those who never use ec2 auth mechanism in future, the fields in role tag might be hard to reason about.

I completely agree. The same holds true with other bindings in the role, and I think that's the hardest part to reason about; role tags seem much easier to me. That's why I originally suggested having these as two entirely different auth backends :)

> If the role has a bound by IAM principal ARN and when the authentication is not performed using identity document, the role tag fields does not represent the subset of capabilities on the role.

I would say that it _partially_ represents the subset of the capabilities on the role (i.e., binding by individual instance ID, specific policies, or `max_ttl`).

> 2 .When the role is bound by IAM principal ARN, don't allow role tags to be created at all.

My concern here is about people changing allowed auth types on the role. Do we just make that immutable on the role? If we don't make it immutable, then do we just delete all the role tags on the role when switching from ec2 to iam? It also makes the transition path more difficult for customers, as they'll potentially have to map one role with role tags specifying policies to many roles with different policies mapped for each role.

> For example, when the auth types contain both iam and ec2, which credentials do we prioritize if both identity document and the new ones are supplied?

We could just set an auth parameter of `auth_type` and ask customers ;) On an actual note, I think we should just error out in that scenario.